br/pkg/restore/internal/prealloc_table_id/alloc.go

```diff
@@ -34,45 +37,126 @@ type PreallocIDs struct {
 	start          int64
 	reusableBorder int64
 	end            int64
-	count          int64
-	next           atomic.Int64
+	hash           [32]byte
+	unallocedIDs   []int64
+	allocRule      map[int64]int64
+}
+
+func NewAndPrealloc(tables []*metautil.Table, m Allocator) (*PreallocIDs, error) {
+	if len(tables) == 0 {
+		return &PreallocIDs{
+			start: math.MaxInt64,
+		}, nil
+	}
+	preallocIDs, err := New(tables)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create preallocIDs")
+	}
+	if err := preallocIDs.PreallocIDs(m); err != nil {
+		return nil, errors.Wrap(err, "failed to allocate prealloc IDs")
+	}
+	return preallocIDs, nil
 }
 
 // New collects the requirement of prealloc IDs and return a
 // not-yet-allocated PreallocIDs.
-func New(tables []*metautil.Table) *PreallocIDs {
+func New(tables []*metautil.Table) (*PreallocIDs, error) {
 	if len(tables) == 0 {
 		return &PreallocIDs{
 			start: math.MaxInt64,
-		}
+		}, nil
 	}
 
 	maxID := int64(0)
-	count := int64(len(tables))
+	unallocedIDs := make([]int64, 0, len(tables))
 
 	for _, t := range tables {
 		if t.Info.ID > maxID && t.Info.ID < InsaneTableIDThreshold {
 			maxID = t.Info.ID
 		}
+		unallocedIDs = append(unallocedIDs, t.Info.ID)
 
 		if t.Info.Partition != nil && t.Info.Partition.Definitions != nil {
-			count += int64(len(t.Info.Partition.Definitions))
 			for _, part := range t.Info.Partition.Definitions {
 				if part.ID > maxID && part.ID < InsaneTableIDThreshold {
 					maxID = part.ID
 				}
+				unallocedIDs = append(unallocedIDs, part.ID)
 			}
 		}
 	}
-	if maxID+count+1 > InsaneTableIDThreshold {
-		return nil
+	if maxID+int64(len(unallocedIDs))+1 > InsaneTableIDThreshold {
+		return nil, errors.Errorf("table ID %d is too large", maxID)
 	}
+
+	sort.Slice(unallocedIDs, func(i, j int) bool { return unallocedIDs[i] < unallocedIDs[j] })
+
 	return &PreallocIDs{
 		start:          math.MaxInt64,
 		reusableBorder: maxID + 1,
-		count:          count,
-		next:           atomic.Int64{},
+		hash:           computeSortedIDsHash(unallocedIDs),
+		unallocedIDs:   unallocedIDs,
+		allocRule:      make(map[int64]int64, len(unallocedIDs)),
+	}, nil
+}
+
+func ReuseIDs(legacy *checkpoint.PreallocIDs, tables []*metautil.Table) (*PreallocIDs, error) {
+	if legacy == nil {
+		return nil, errors.Errorf("no prealloc IDs to be reused")
+	}
+
+	maxID := int64(0)
+	ids := make([]int64, 0, len(tables))
+	for _, t := range tables {
+		if t.Info.ID > maxID && t.Info.ID < InsaneTableIDThreshold {
+			maxID = t.Info.ID
+		}
+		ids = append(ids, t.Info.ID)
+
+		if t.Info.Partition != nil && t.Info.Partition.Definitions != nil {
+			for _, part := range t.Info.Partition.Definitions {
+				if part.ID > maxID && part.ID < InsaneTableIDThreshold {
+					maxID = part.ID
+				}
+				ids = append(ids, part.ID)
+			}
+		}
+	}
```

Comment:
This logic is almost the same as `NewAndAlloc`, would you extract it to a function?